File: java/completed/euler0027.java
import lib.EulerTime;
import lib.EulerUtil;

import java.util.ArrayList;

/**
 * Created by Holden on 12/19/2016. - COMPLETED
 *
 * Problem 27 - Quadratic Primes
 *
 */

public class Prob0027 {
    public static void main(String[] args) {
        long startTime = System.nanoTime();
        int answer = quadraticPrimes(1000);
        long endTime = System.nanoTime();

        System.out.println(answer);

        long duration = (endTime - startTime);
        EulerTime.time(duration);
    }

    public static int quadraticPrimes(int bound) {
        int maxPrimes = 1;
        int maxa = 1;
        int maxb = 1;

        for (int a = -(bound-1); a < bound; a++) {
            for (int b = -bound; b <= bound; b++) {
                if (b % 2 == 0) {
                    continue;
                }

                ArrayList<Integer> primes = new ArrayList<>();
                int check = 0;
                long quadCheck = quadratic(a, b, check);

                while (quadCheck > 0 && EulerUtil.isPrime(quadCheck)) {
                    primes.add(check);
                    check++;
                    quadCheck = quadratic(a, b, check);
                }

                if (primes.size() > maxPrimes) {
                    maxPrimes = primes.size();
                    maxa = a;
                    maxb = b;
                }
            }
        }

        return maxa * maxb;
    }

    private static long quadratic(int a, int b, int n) {
        return EulerUtil.pow(n, 2) + (n * a) + b;
    }
}

File: javascript/completed/euler0027.js
import { isPrime as _isPrime } from '../util.js';

function isPrime(n, primeMap) {
  if (primeMap.get(n) === undefined) {
    const res = _isPrime(n);
    primeMap.set(n, res);
  }
  return primeMap.get(n);
}

function quadratic(a, b, x) {
  return x * x + a * x + b;
}

export async function solve27() {
  const n = 1000;
  let maxStreak = 0;
  let maxStreakCofs = { a: 0, b: 0 };
  let primeMap = new Map();
  for (let b = 1; b <= n; b++) {
    if (isPrime(b, primeMap)) {
      for (let a = 1; a < n; a++) {
        let x = 0;
        let result = b;
        while (isPrime(result, primeMap)) {
          x++;
          result = quadratic(a, b, x);
        }
        if (x > maxStreak) {
          maxStreak = x;
          maxStreakCofs = { a, b };
        }
      }
      for (let a = -1; a > -n; a--) {
        let x = 0;
        let result = b;
        while (isPrime(result, primeMap) && result > 0) {
          x++;
          result = quadratic(a, b, x);
        }
        if (x > maxStreak) {
          maxStreak = x;
          maxStreakCofs = { a, b };
        }
      }
    }
  }

  return maxStreakCofs.a * maxStreakCofs.b;
}
